### PR TITLE
Issue #566: Add metafunctions and prop to deduce common value_type

### PR DESCRIPTION
**Pull request by ndellingwood**
Addresses issue #566 to provide functionality in Kokkos to
replace ViewFactory in Sacado

### CODE REVIEW

**Comment by ndellingwood** [comment]
@hcedwar For review; I still need to run testing through test_all_sandia, though this commit should not affect anything else within Kokkos.

**Comment by ndellingwood** [comment]
@hcedwar @etphipp One definite item for consideration is naming of the user function, e.g. `common_view_alloc_prop`, `common_view_traits`, ...

File: core/unit_test/TestViewCtorPropEmbeddedDim.hpp [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    using VIT = typename TestViewCtorProp_EmbeddedDim::ViewIntType ;
+    using VDT = typename TestViewCtorProp_EmbeddedDim::ViewDoubleType ;
+
+    VIT vi1("vi1", N0, N1);
+    VDT vd1("vd1", N0);
+
+    // TEST: Test for common type between two views, one with type double, other with type int
+    // Deduce common value_type and construct a view with that type
+    using CommonViewValueType = typename decltype( Kokkos::common_view_alloc_prop(vi1, vd1))::value_type;
+    auto view_alloc_arg=  Kokkos::common_view_alloc_prop(vi1, vd1);
```
Comment (hcedwar) [inline core/unit_test/TestViewCtorPropEmbeddedDim.hpp]: For the example, would the following ordering work?  A simpler pattern if it works.
```c++
auto view_alloc_arg = ...etc. ;
using CommonViewValueType = typename decltype( view_alloc_arg )::value_type ;
```

File: core/src/Kokkos_View.hpp [hunk 2]
```diff
@@ -357,11 +357,123 @@ struct ViewTraits {
...
+template< class D, class ... P >
+struct is_view< const View<D,P...> > : public std::true_type {};
+
+
 } /* namespace Kokkos */
 
 //----------------------------------------------------------------------------
 //----------------------------------------------------------------------------
 
+namespace Kokkos { namespace Impl {
```
Comment (hcedwar) [inline core/src/Kokkos_View.hpp]: This section should appear after the View implementation, and then would not have to move the `is_view` fragment as well.

File: core/src/Kokkos_View.hpp [hunk 3]
```diff
@@ -357,11 +357,123 @@ struct ViewTraits {
...
+
+
+template < class Specialize, class ValueType >
+struct CommonViewAllocProp;
+
+template < class ValueType >
+struct CommonViewAllocProp< void, ValueType >
+{
+  using value_type = ValueType;
+//  enum : bool { is_view = IsView };
```
Comment (hcedwar) [inline core/src/Kokkos_View.hpp]: If not needed then delete.

File: core/src/Kokkos_View.hpp [hunk 4]
```diff
@@ -357,11 +357,123 @@ struct ViewTraits {
...
+template < class FirstView >
+struct DeduceCommonViewAllocProp< FirstView >
+{
+  using specialize = typename FirstView::traits::specialize;
+
+  using value_type = typename FirstView::traits::value_type;
+
+  enum : bool { is_view = is_view< FirstView >::value };
+
+  //using prop_type = CommonViewAllocProp< specialize, value_type, is_view >; // Issue: No way to know if IsView for casting within ViewCtorProp
```
Comment (hcedwar) [inline core/src/Kokkos_View.hpp]: If should not be here then delete.

**Comment by ndellingwood** [comment]
@hcedwar I pushed changes to address the requests in your review: 
* cleanup stray comments
* improve unit test use pattern of the new feature
* move new stuff in Kokkos_View to bottom of the file

**Comment by ndellingwood** [comment]
@hcedwar I ran unit tests with the various backends primarily with a generated makefile. Sparse spot-checking with test_all_sandia looks good as well.

**Comment by ndellingwood** [comment]
@hcedwar @etphipp if the commit looks good technically at this point, the final concern I have is naming. Is `common_view_alloc_prop` acceptable? Another possible name was `common_view_traits` ... if something else is preferred please let me know and I'll make the changes promptly.

**Pull request merged**